chill-scala/src/test/scala/com/twitter/chill/ClosureCleanerSpec.scala

```diff
@@ -16,59 +16,227 @@ limitations under the License.
-        x * 2
+
+    "clean basic serializable closures" in {
+      val localVal = 1
+      val closure1 = (_: Int) => localVal
+      val closure2 = (x: Int) => x + localVal
+
+      serializableFn(closure1, before = true, after = true)
+      serializableFn(closure2, before = true, after = true)
+    }
+
+    "clean basic closures (non LMF)" in {
+      assume(!ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (_: Int) => someSerializableValue
+      val closure2 = (_: Int) => someSerializableMethod()
+
+      serializableFn(closure1, before = false, after = true)
+      serializableFn(closure2, before = false, after = true)
+    }
+
+    "clean basic closures (LMF)" in {
+      assume(ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (_: Int) => someSerializableValue
+      val closure2 = (_: Int) => someSerializableMethod()
+
+      serializableFn(closure1, before = false, after = false)
+      serializableFn(closure2, before = false, after = false)
+    }
+
+    "clean functions in traits" in {
+      // not serializable; AwesomeFn2 needs to extend Serializable
+      val fn = BaseFns2.timesByMult
+      serializableFn(fn, before = false, after = false)
+    }
+
+    "clean basic nested closures (non LMF)" in {
+      assume(!ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (i: Int) => {
+        Option(i).map { x =>
+          x + someSerializableValue
+        }
+      }
+      val closure2 = (j: Int) => {
+        Option(j).map { x =>
+          x + someSerializableMethod()
+        }
+      }
+      val closure3 = (m: Int) => {
+        Option(m).foreach { x =>
+          Option(x).foreach { y =>
+            Option(y).foreach { z =>
+              someSerializableValue
+            }
+          }
+        }
       }
 
-      ClosureCleaner(myFun)
-      myFun(1) should equal(2)
-      myFun(2) should equal(4)
+      serializableFn(closure1, before = false, after = true)
+      serializableFn(closure2, before = false, after = true)
+      serializableFn(closure3, before = false, after = true)
+    }
 
-      case class Test(x: Int)
-      val t = Test(3)
-      ClosureCleaner(t)
-      t should equal(Test(3))
+    "clean basic nested closures (LMF)" in {
+      assume(ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (i: Int) => {
+        Option(i).map { x =>
+          x + someSerializableValue
+        }
+      }
+      val closure2 = (j: Int) => {
+        Option(j).map { x =>
+          x + someSerializableMethod()
+        }
+      }
+      val closure3 = (m: Int) => {
+        Option(m).foreach { x =>
+          Option(x).foreach { y =>
+            Option(y).foreach { z =>
+              someSerializableValue
+            }
+          }
+        }
+      }
+
+      serializableFn(closure1, before = false, after = false)
+      serializableFn(closure2, before = false, after = false)
+      serializableFn(closure3, before = false, after = false)
```

Comment:
we can't support these cases? What is the goal of this test? Why can't these cases be supported, can you add a comment?

Reply:
This is an interesting test ... we have 2 tests with the same thing; closures reference properties in ClosureCleanerSpec; one test is non-LMF and ther other LMF. However under 2.11 (non-LMF) we have access to ther $outer ref and we can clean it but with 2.12 LMF Lambda has no outer ref; I couldn't find a way to clean these cases yet.

Reply:
okay, in any case, we seem to have made progress.

Reply:
Yeah we made some progress here. As soon as we release this I can actually use it in https://github.com/spotify/scio